src/SpectralDataManager.cpp

```diff
@@ -0,0 +1,214 @@
+   // The calculated frequency peak will be stored in mReturnFreq
+   if (!TrackSpectrumTransformer::Process( SnappingProcessor, wt,
+                                           1, startSC, 2048))
+      return 0;
+
+   return mSnapReturnFreq;
+}
+
+bool SpectralDataManager::Worker::SnappingProcessor(SpectrumTransformer &transformer) {
+   auto &worker = static_cast<Worker &>(transformer);
+   // Compute power spectrum in the newest window
+   {
+      MyWindow &record = worker.NthWindow(0);
+      float *pSpectrum = &record.mSpectrums[0];
+      const double dc = record.mRealFFTs[0];
+      *pSpectrum++ = dc * dc;
+      float *pReal = &record.mRealFFTs[1], *pImag = &record.mImagFFTs[1];
+      for (size_t nn = worker.mSpectrumSize - 2; nn--;) {
+         const double re = *pReal++, im = *pImag++;
+         *pSpectrum++ = re * re + im * im;
+      }
+      const double nyquist = record.mImagFFTs[0];
+      *pSpectrum = nyquist * nyquist;
+
+      const double &sr = worker.mSnapSamplingRate;
+      const double nyquistRate = sr / 2;
+      const double &threshold = worker.mSnapThreshold;
+      const double &spectrumSize = worker.mSpectrumSize;
+      const wxInt64 &targetFreq = worker.mSnapTargetFreq;
+
+      int targetBin = targetFreq / nyquistRate * spectrumSize;
+      int binBound = spectrumSize * threshold;
+      float maxValue = std::numeric_limits<float>::min();
+
+      // Skip the first and last bin
+      for(int i = -binBound; i < binBound; i++){
+         int idx = targetBin + i;
+         if(record.mSpectrums[idx] > maxValue){
+            maxValue = record.mSpectrums[idx];
+            // Update the return frequency
+            worker.mSnapReturnFreq = idx / spectrumSize * nyquistRate;
+         }
+      }
+   }
+
+   return true;
+}
+
+bool SpectralDataManager::Worker::Processor(SpectrumTransformer &transformer)
+{
+   auto &worker = static_cast<Worker &>(transformer);
+   // Compute power spectrum in the newest window
+   {
+      MyWindow &record = worker.NthWindow(0);
+      float *pSpectrum = &record.mSpectrums[0];
+      const double dc = record.mRealFFTs[0];
+      *pSpectrum++ = dc * dc;
+      float *pReal = &record.mRealFFTs[1], *pImag = &record.mImagFFTs[1];
+      for (size_t nn = worker.mSpectrumSize - 2; nn--;) {
+         const double re = *pReal++, im = *pImag++;
+         *pSpectrum++ = re * re + im * im;
+      }
+      const double nyquist = record.mImagFFTs[0];
+      *pSpectrum = nyquist * nyquist;
+   }
+
+   worker.ApplyEffectToSelection();
+   return true;
+}
+
+bool SpectralDataManager::Worker::ApplyEffectToSelection() {
+   auto &record = NthWindow(0);
+   auto nyquist = mpSpectralData->GetSR() / 2;
+
+   for(const auto &spectralDataMap: mpSpectralData->dataHistory){
+      for(const auto &data: spectralDataMap){
+         long long sc = data.first;
+         // Check if the data's time sampleCount within this window
+         if(sc >= mStartSample && sc < mEndSample){
```

Comment:
How will this change if all sample counts are rounded to hop sizes?  Will it mean that the inner for-loop needs only one pass?